### PR TITLE
Do not generate dummy code to force module linking

### PR DESCRIPTION
**Pull request by chambart**
To ensure that a module is linked if only C externals are used from its interface, there is some dummy code generated referencing its top-level module.
This avoids it by propagating separately the required modules and registering it when needed.

This generated code is problematic for some other ongoing development.

### CODE REVIEW

File: asmcomp/asmpackager.ml [hunk 1]
```diff
@@ -110,8 +110,11 @@ let make_package_object ppf members targetobj targetname coercion
         ~module_ident
         ~module_initializer:lam
     in
+    let required_globals =
+      failwith "TODO: fill required_globals"
```
Comment (trefis) [inline asmcomp/asmpackager.ml]: I'm guessing you're missing something here.

Comment (chambart) [inline asmcomp/asmpackager.ml]: Oups...
Who cares about pack anywhay ? ;-)

Comment (chambart) [inline asmcomp/asmpackager.ml]: In fact there is nothing to do in case of a pack

**Comment by garrigue** [comment]
Good idea. This generated code was a hack anyway, I just could not find a simple way to do it.
(The number of modified files shows that this is not completely trivial.)
Have you seen whether this can be done for bytecode too?
But it's probably less problematic there.

**Comment by chambart** [comment]
The change is not so large, it's mainly interface propagation. And 8 files changes are test files.
I will take a look at bytecode and closure to see if it can be made a cleaner there too easily.
I have no idea what caused the appveyor failure, there are no log to help...

**Comment by chambart** [comment]
And here it is for bytecode and closure. The changes required for the bytecode version are a bit more invasive than the native one since there was already a field in cmx to record the information, but none for bytecode.

**Comment by chambart** [comment]
I forgot to bootstrap, hence the build failure.

@gasche In this case if we keep the blob out of the real change commit, there is no way to get a working intermediate state, hence breaking bisection. Would you prefer a squashed version of the patch containing both the change and the bootstrap here ?

**Comment by gasche** [comment]
I think squashing would be fine if there is a well-defined commit in which to squash. Sometimes we do changes iteratively in ways that actually require doing several bootstraps in the intermediate states, and then it can be a bit cumbersome to completely preserve bisectability.

**Comment by chambart** [comment]
@bobot The cmo format changed. This should probably not require bootstraping.

The build without bootstrap is failing when building otherlibs which use the newly built compiler instead of the `boot/ocamlc` one, see: https://travis-ci.org/chambart/ocaml-1/jobs/134999150

**Comment by chambart** [comment]
Rebased to clean the history and make bisectable.

**Comment by chambart** [comment]
Rebased, there was some changes needed to merge with the recent location changes.

@garrigue, @xavierleroy, any objection ?

**Comment by garrigue** [comment]
This looks ok to me.
The obvious difficulty is that this kind of things are hard to test, but I hope you've done enough checks.

**Comment by chambart** [comment]
I tried some complex cases and built some subset of opam. In fact opam testing does not convince me that much, I don't think that exercise this a lot.

**Comment by chambart** [comment]
Rebased

**Pull request merged**